File: rd-framework/src/commonMain/kotlin/com/jetbrains/rd/framework/Protocol.kt
package com.jetbrains.rd.framework

import com.jetbrains.rd.framework.base.RdExtBase
import com.jetbrains.rd.framework.impl.InternRoot
import com.jetbrains.rd.util.getLogger
import com.jetbrains.rd.util.lifetime.Lifetime
import com.jetbrains.rd.util.reactive.IScheduler
import com.jetbrains.rd.util.reactive.ViewableSet
import com.jetbrains.rd.util.string.RName

class Protocol(
        override val serializers: ISerializers,
        override val identity: IIdentities,
        override val scheduler: IScheduler,
        override val wire: IWire, //to initialize field with circular dependencies
        val lifetime: Lifetime,
        serializationCtx: SerializationCtx? = null
) : IRdDynamic, IProtocol {

    override val location: RName = RName.Empty
    override val outOfSyncModels: ViewableSet<RdExtBase> = ViewableSet()

    companion object {
        val logCategory = "protocol"
        fun sublogger(subcategory: String) = getLogger("$logCategory.$subcategory")
        val initializationLogger = sublogger("INIT")
    }

    override val protocol: IProtocol get() = this
    override val serializationContext: SerializationCtx = serializationCtx ?: SerializationCtx(serializers, mapOf("Protocol" to InternRoot().also {
        it.rdid = RdId.Null.mix("ProtocolInternRoot")
        scheduler.queue {
            it.bind(lifetime, this, "ProtocolInternRoot")
        }
    }))
}

File: rd-cpp/src/rd_framework_cpp/src/main/Protocol.cpp
#include "Protocol.h"

#include "SerializationCtx.h"
#include "InternRoot.h"

#include <utility>

namespace rd {
	const Logger Protocol::initializationLogger;

	void Protocol::initialize() const {
		internRoot = std::make_unique<InternRoot>();
		
		context = std::make_unique<SerializationCtx>(serializers.get(), SerializationCtx::roots_t{{util::getPlatformIndependentHash("Protocol"), internRoot.get()}});


		internRoot->rdid = RdId::Null().mix(InternRootName);
		scheduler->queue([this] {
			internRoot->bind(lifetime, this, InternRootName);
		});
	}

	Protocol::Protocol(std::shared_ptr<Identities> identity, IScheduler *scheduler, std::shared_ptr<IWire> wire, Lifetime lifetime) :
			IProtocol(std::move(identity), scheduler, std::move(wire)), lifetime(lifetime) {
		// initialize();
	}

	Protocol::Protocol(Identities::IdKind kind, IScheduler *scheduler, std::shared_ptr<IWire> wire, Lifetime lifetime) :
			IProtocol(std::make_shared<Identities>(kind), scheduler, std::move(wire)), lifetime(lifetime) {
		// initialize();
	}

	Protocol::~Protocol() = default;

	const SerializationCtx &Protocol::get_serialization_context() const {
		if (!context) {
			initialize();
		}
		return *context;
	}

}
